staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -1439,6 +1441,24 @@ func (c *Cacher) Wait(ctx context.Context) error {
 	return c.ready.wait(ctx)
 }
 
+// setInitialEventsEndBookmarkIfRequested sets initialEventsEndBookmark field in watchCacheInterval for watchlist request
+func (c *Cacher) setInitialEventsEndBookmarkIfRequested(cacheInterval *watchCacheInterval, opts storage.ListOptions) {
+	if opts.SendInitialEvents != nil && *opts.SendInitialEvents && opts.Predicate.AllowWatchBookmarks {
+		initialEventsEndBookmark := &watchCacheEvent{
+			Type:            watch.Bookmark,
+			Object:          c.newFunc(),
+			ResourceVersion: c.watchCache.resourceVersion,
+		}
+
+		if err := c.versioner.UpdateObject(initialEventsEndBookmark.Object, initialEventsEndBookmark.ResourceVersion); err != nil {
+			klog.Errorf("failure to set resourceVersion to %d on initialEventsEndBookmark event %+v for watchlist request and wait for bookmark trigger to send", initialEventsEndBookmark.ResourceVersion, initialEventsEndBookmark.Object)
+			initialEventsEndBookmark = nil
+		}
+
+		cacheInterval.initialEventsEndBookmark = initialEventsEndBookmark
```

Comment:
maybe adjust the order would be more clear.

```suggestion
		object := c.newFunc()
		if err := c.versioner.UpdateObject(object, initialEventsEndBookmark.ResourceVersion); err != nil {
			klog.Errorf("failure to set resourceVersion to %d on initialEventsEndBookmark event %+v for watchlist request and wait for bookmark trigger to send", initialEventsEndBookmark.ResourceVersion, initialEventsEndBookmark.Object)
			return
		}

		cacheInterval.initialEventsEndBookmark = &watchCacheEvent{
			Type:            watch.Bookmark,
			Object:          object,
			ResourceVersion: c.watchCache.resourceVersion,
		}
```

Reply:
Thanks a lot for review. I'm just trying to be consistent with bookmarkEvent generation in `dispatchEvents` method. I don't know whether it is worthy to modify. lol

Reply:
this is equivalent - I don't see a difference, so no need to change :)

Reply:
well, If both of two places shares same logic, maybe extra it as a function/method would be better, instead copy&paste to another place and try to be consistent with both side?

https://github.com/kubernetes/kubernetes/blob/d495dbc56a4dcdbfcc5e7b88822f02daa94214c5/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L1073-L1081